lom/src/lib/lomcommon/helper.go

```diff
@@ -151,6 +152,100 @@ func LogDebug(s string, a ...interface{})  {
     LogMessage(syslog.LOG_DEBUG, s, a...)
 }
 
+/*
+ * System Shutdown
+ *
+ */
+type sysShutdown_t struct {
+    ch          chan int
+    wg          *sync.WaitGroup
+    shutdown    bool
```

Comment:
shutdown must be protected by lock ?

Reply:
Answered above.

Reply:
if register()  and doShutdown() started at same time ? Then shutdown may be accessed parallelly ?

Reply:
1. register is skipped if shutdown flag set.
2. Even if that is escaped in corner cases, syncgroup handle all concurrency. One increment & other wait on routine safe object,

Reply:
if !p.shutdown in register()     and    p.shutdown = true in doShutdown()  executing at same time is race condition. Practically syncgroup may handle it, but logically I'm thinking its issue ??

Reply:
This is precisely, my point number 2.
SyncGroup would handle it.

Worst possible scenario, that we may not wait for the last sneaked in routine. This is completely OK as process is exiting.

All above are for theoretical analysis only. Practical probability is  nearly 0  as all the clients of this API are long running processes that kick off nearly at the start of process and hence close to shutdown, the probability of registrations is near nil.